File: Sources/S3mailKit/Compose.swift
// SPDX-FileCopyrightText: 2026 Kai Ole Hartwig <mail@ole-hartwig.eu>
// SPDX-License-Identifier: Apache-2.0

import Foundation

/// What somebody has written, on its way to the core.
///
/// The field names are the core's, not this file's: `mailer.Draft` on the Go
/// side reads exactly these keys, and the desktop's compose window sends the
/// same shape. One vocabulary for a mail being written, and neither end has to
/// translate.
public struct Draft: Codable, Equatable, Identifiable {

    /// Where this draft came from - and therefore what the core does with the
    /// original: quote it, attach it, or nothing at all.
    public enum Mode: String, Codable {
        case new, reply, forward
    }

    public var mode: Mode = .new
    /// The message being replied to or forwarded. Empty for a new mail.
    public var key: String = ""
    /// Empty means the mailbox's own sender address. The phone rarely sets it.
    public var from: String = ""
    public var to: String = ""
    public var cc: String = ""
    public var bcc: String = ""
    public var subject: String = ""
    public var body: String = ""
    /// The draft this replaces. Set by `save()` so the next save overwrites
    /// rather than leaving a trail of half-written mails in the folder.
    public var draftKey: String = ""
    public var attachments: [Attachment] = []

    public struct Attachment: Codable, Equatable, Identifiable {
        public let filename: String
        public let contentType: String
        /// Base64 on the wire in both directions - that is what Go's
        /// encoding/json does with []byte, so neither side has to say so.
        public let content: Data
        public var id: String { filename }

        public init(filename: String, contentType: String, content: Data) {
            self.filename = filename
            self.contentType = contentType
            self.content = content
        }

        enum CodingKeys: String, CodingKey {
            case filename, content
            case contentType = "content_type"
        }
    }

    enum CodingKeys: String, CodingKey {
        case mode, key, from, to, cc, bcc, subject, body, attachments
        case draftKey = "draft_key"
    }

    /// Identity for the sheet that presents it. Not part of the wire format -
    /// the core never sees it, and two drafts of the same mail are two windows.
    public let id = UUID()

    public init() {}

    /// A reply, with the quoting left to the core. The subject prefix and the
    /// threading headers are its job too - this only says which message, which
    /// way, and to whom.
    ///
    /// The recipient comes from Reply-To when there is one. Mailing lists and
    /// ticket systems set it, and an answer that ignores it lands with whoever
    /// pressed send instead of with the list.
    ///
    /// Reply-all puts the original To and Cc into Cc - **including your own
    /// address**, and that is on purpose: it is exactly what the desktop does
    /// (`compose(kind, all)` in inbox.html). Two clients on the same mailbox
    /// behaving differently is worse than either behaviour, and if this should
    /// change it should change in both places.
    public static func replying(to full: Mailbox.Full, key: String,
                                all: Bool = false) -> Draft {
        var d = Draft()
        d.mode = .reply
        d.key = key
        d.to = full.replyTo.isEmpty ? full.from : full.replyTo
        if all {
            d.cc = [full.to, full.cc].filter { !$0.isEmpty }.joined(separator: ", ")
        }
        return d
    }

    /// A forward. No recipient - somebody has to say where it goes, and a
    /// prefilled one is how a mail ends up with the wrong person.
    public static func forwarding(_ key: String) -> Draft {
        var d = Draft()
        d.mode = .forward
        d.key = key
        return d
    }

    /// Whether this is worth storing. An empty draft saved on every keystroke
    /// would put an object in the bucket for a window somebody opened and shut.
    public var isWorthKeeping: Bool {
        !(to.isEmpty && cc.isEmpty && bcc.isEmpty
          && subject.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
          && body.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
          && attachments.isEmpty)
    }
}

/// What the core answers when something could not be done.
///
/// The bridge sends a code, never a sentence: a Go library inside an app has no
/// business holding text in one language while the system is set to another.
/// The sentences are here, and they come from the app's own catalogue.
public enum ComposeError: String, Error, LocalizedError {
    case noRecipient = "no_recipient"
    case noSender = "no_sender"
    case tooLarge = "too_large"
    /// The device's key may read the bucket but not send. That is the wizard's
    /// default, and it is the right one for a phone that only reads.
    case notPermitted = "send_not_permitted"
    case refused = "send_refused"
    case badDraft = "bad_draft"

    public var errorDescription: String? {
        NSLocalizedString("compose.\(rawValue)", bundle: .module,
                          comment: "why a mail could not be sent")
    }

    /// Anything the bridge did not name stays an NSError and is shown as it
    /// came. An unknown error said plainly is more use than a familiar one said
    /// wrongly.
    static func from(_ error: Error) -> Error {
        ComposeError(rawValue: (error as NSError).localizedDescription) ?? error
    }
}

/// What happened after SES took the message.
///
/// `warning` is the part that matters: from the moment SES answers, the mail is
/// out of the house, and a copy that did not reach sent/ must never be reported
/// as a failure to send. Somebody reading "failed" sends it again.
public struct SendResult: Decodable, Equatable {
    public let messageID: String
    public let warning: Warning?

    public enum Warning: String, Decodable {
        case notRecorded = "send_not_recorded"
        case sentNotStored = "sent_not_stored"
        case draftNotRemoved = "draft_not_removed"

        public var text: String {
            NSLocalizedString("compose.\(rawValue)", bundle: .module,
                              comment: "the mail went out, but something after it did not")
        }
    }

    enum CodingKeys: String, CodingKey {
        case messageID = "message_id"
        case warning
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        messageID = try c.decodeIfPresent(String.self, forKey: .messageID) ?? ""
        warning = try? c.decodeIfPresent(Warning.self, forKey: .warning)
    }
}

/// A send that started and whose end nobody witnessed.
///
/// The app asks about these on every start. It has to ask rather than decide:
/// guessing "sent" loses a mail, guessing "not sent" sends it twice, and only
/// the person can look in their own sent folder and say which it was.
public struct PendingSend: Decodable, Identifiable, Equatable {
    public let key: String
    public let to: String
    public let subject: String
    public let started: String
    public var id: String { key }

    enum CodingKeys: String, CodingKey { case key, to, subject, started }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        key = try c.decode(String.self, forKey: .key)
        to = try c.decodeIfPresent(String.self, forKey: .to) ?? ""
        subject = try c.decodeIfPresent(String.self, forKey: .subject) ?? ""
        started = try c.decodeIfPresent(String.self, forKey: .started) ?? ""
    }
}

/// The text catalogue, reachable from outside for the test that checks it.
///
/// The desktop has i18n_test.go: every language carries every key, and the
/// placeholders match. Without that, a language falls behind silently - the
/// missing sentence is simply the key, and nobody sees it until somebody
/// switches their phone to Spanish.
public enum Catalogue {
    public static let bundle = Bundle.module
    public static let languages = ["de", "en", "es"]
}

File: mobile/compose.go
// SPDX-FileCopyrightText: 2026 Kai Ole Hartwig <mail@ole-hartwig.eu>
// SPDX-License-Identifier: Apache-2.0

package mobile

import (
	"encoding/json"
	"errors"
	"strings"
	"time"

	"github.com/ohartwig/s3mail/awsx"
	"github.com/ohartwig/s3mail/core"
	"github.com/ohartwig/s3mail/mailer"
	"github.com/ohartwig/s3mail/mimeparse"
	"github.com/ohartwig/s3mail/store"
)

// Writing and sending, from the phone.
//
// The order of operations here is not this file's invention - it is the one
// web/send.go arrived at, and it is copied deliberately rather than improved
// upon. Two rules make it what it is:
//
//   - Write down what is about to happen before it happens. The marker in
//     store/sending.go survives a process that dies mid-send, which on a phone
//     is not an edge case: iOS kills apps that go to the background at the
//     wrong moment, and it does so without asking.
//   - Once SES has taken the message, nothing afterwards may turn the answer
//     into a failure. A copy that did not reach sent/ is a warning. Reported as
//     an error it reads as "not sent", and the next thing that happens is the
//     same mail going out twice.
//
// The phone makes the first rule matter more and the second one no less.

// Errors here are codes, not sentences. Every user-visible string on this
// platform lives in Swift, where the system's own language settings decide -
// a Go library inside an app has no business holding German text. So the
// bridge names the case and Swift says it out loud.
var (
	errNoRecipient = errors.New("no_recipient")
	errNoSender    = errors.New("no_sender")
	errTooLarge    = errors.New("too_large")
	errNotAllowed  = errors.New("send_not_permitted")
	errRefused     = errors.New("send_refused")
)

// sender is built on demand, not in Open.
//
// A device may hold a key that is allowed to read the bucket and nothing else -
// that is the default the wizard hands out, and for a phone that only reads it
// is the right one. Building an SES client for such a device on every start
// would cost a round trip to discover what the policy already says: nothing.
func (m *Mailbox) ses() *awsx.SES {
	if m.sender == nil {
		m.sender = awsx.NewSES(m.cfg, "")
	}
	return m.sender
}

// SaveDraft stores what somebody has written, without sending it.
//
// The draft is a real message in drafts/ - not a record in some app-local
// store. That is what makes it visible from the desk five minutes later, and it
// is why the phone needs no synchronisation of its own for the one thing people
// most expect to follow them between devices.
func (m *Mailbox) SaveDraft(draftJSON string) (string, error) {
	e, err := m.draft(draftJSON)
	if err != nil {
		return "", err
	}
	n, err := mailer.BuildDraft(e, m.setup.From, mailer.Original{}, time.Now())
	if err != nil {
		return "", code(err)
	}
	// The name comes from the Message-ID, so saving twice replaces rather than
	// piles up - which on a phone happens on every autosave.
	key, err := m.inner.Put(ctx(), core.Drafts, n.ID, n.Raw)
	if err != nil {
		return "", err
	}
	// The old draft only goes once the new one is safely stored. A draft the
	// core renamed is still the same draft; deleting first would lose it if the
	// write below failed.
	if e.DraftKey != "" && e.DraftKey != key {
		_ = m.inner.DropDraft(ctx(), e.DraftKey)
	}
	return asJSON(map[string]any{"key": key, "message_id": n.ID})
}

// DropDraft throws a draft away. Deliberately its own call and not a flag on
// SaveDraft: discarding is a decision, and it should read like one.
func (m *Mailbox) DropDraft(key string) error {
	return m.inner.DropDraft(ctx(), key)
}

// Send builds the message, records the attempt, hands it to SES, and files the
// copy. The sequence is web/send.go's; see the comment at the top of this file
// for why it may not be rearranged.
func (m *Mailbox) Send(draftJSON string) (string, error) {
	e, err := m.draft(draftJSON)
	if err != nil {
		return "", err
	}
	if m.setup.From == "" {
		return "", errNoSender
	}

	var o mailer.Original
	if e.Key != "" {
		// A reply needs the original's Message-ID and References or the thread
		// falls apart in everybody else's client; a forward needs the whole
		// message, because that is what gets attached.
		if raw, err := m.inner.Fetch(ctx(), e.Key, 0); err == nil {
			full := mimeparse.Read(raw, time.Unix(0, 0).UTC())
			o = mailer.Original{MessageID: full.MessageID,
				References: full.References, Subject: full.Subject}
			if e.Mode == "forward" {
				o.Raw = raw
			}
		}
	}

	n, err := mailer.Build(e, m.setup.From, o, time.Now())
	if err != nil {
		return "", code(err)
	}

	// A marker that cannot be written does not stop the send: it is a safety
	// net against a crash, not a permission to send. Refusing here would turn a
	// bucket problem into "you cannot send mail", which is the bigger failure.
	marker, markerErr := m.inner.BeginSend(ctx(), store.Sending{
		To: strings.Join(n.To, ", "), Subject: e.Subject, DraftKey: e.DraftKey,
		MessageID: n.ID, Raw: n.Raw})

	id, err := m.ses().Send(ctx(), n)
	if err != nil {
		// Nothing left the house, so the marker would only raise a question
		// with no content behind it.
		if markerErr == nil {
			_ = m.inner.AbandonSend(ctx(), marker)
		}
		if awsx.IsAuthProblem(err) {
			// Almost always the device policy: a key handed out for reading has
			// no ses:SendRawEmail, and this is where that first shows.
			return "", errNotAllowed
		}
		return "", errRefused
	}

	// From here the message is out of the house. Everything below reports
	// through "warning" and never through an error.
	out := map[string]any{"message_id": id}
	if markerErr != nil {
		out["warning"] = "send_not_recorded"
	} else if err := m.inner.MarkSent(ctx(), marker, id); err != nil {
		out["warning"] = "sent_not_stored"
	}
	if _, err := m.inner.Put(ctx(), core.Sent, n.ID, n.Raw); err != nil {
		out["warning"] = "sent_not_stored"
	}
	if e.DraftKey != "" {
		if err := m.inner.DropDraft(ctx(), e.DraftKey); err != nil {
			out["warning"] = "draft_not_removed"
		}
	}
	if markerErr == nil {
		_ = m.inner.AbandonSend(ctx(), marker)
	}
	return asJSON(out)
}

// PendingSends is what the app asks on every start: did anything not finish?
//
// store.RecoverSends closes what it can decide by itself - a message that made
// it to sent/ clearly went out. What comes back is only what nobody but the
// person can answer, and the app has to ask rather than guess. Guessing "sent"
// loses a mail; guessing "not sent" sends it twice.
func (m *Mailbox) PendingSends() (string, error) {
	open, err := m.inner.RecoverSends(ctx())
	if err != nil {
		return "", err
	}
	list := make([]map[string]any, 0, len(open))
	for _, s := range open {
		list = append(list, map[string]any{
			"key": s.Key, "to": s.To, "subject": s.Subject, "started": s.Started})
	}
	return asJSON(list)
}

// ResolveSending records the answer. sent=true files the copy and closes the
// marker; sent=false puts the message back among the drafts.
func (m *Mailbox) ResolveSending(key string, sent bool) error {
	return m.inner.ResolveSending(ctx(), key, sent)
}

// CanSend says whether this device was given a sender address at all. It does
// not ask AWS: the policy is only proven by an actual send, and a check that
// costs a round trip to return a maybe is worth less than the round trip.
func (m *Mailbox) CanSend() bool { return m.setup.From != "" }

// draft parses what Swift sent. Attachment bytes arrive base64-encoded, which
// is what encoding/json does with []byte in both directions - so the two sides
// agree without either of them having to say so.
func (m *Mailbox) draft(draftJSON string) (mailer.Draft, error) {
	var e mailer.Draft
	if err := json.Unmarshal([]byte(draftJSON), &e); err != nil {
		return e, errors.New("bad_draft")
	}
	return e, nil
}

// code turns the mailer's sentinels into the names Swift knows. Anything else
// passes through unchanged: an unknown error said plainly is more use than a
// familiar one said wrongly.
func code(err error) error {
	switch {
	case errors.Is(err, mailer.ErrNoRecipient):
		return errNoRecipient
	case errors.Is(err, mailer.ErrNoSender):
		return errNoSender
	case errors.Is(err, mailer.ErrTooLarge):
		return errTooLarge
	}
	return err
}
